fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java

```diff
@@ -54,7 +54,7 @@ public Substring(Expression str, Expression pos) {
 
     @Override
     public DataType getDataType() {
-        return StringType.INSTANCE;
+        return VarcharType.SYSTEM_DEFAULT;
```

Comment:
could `new VarcharType(len)` works?

Reply:
keep the same type as OriginPlanner, or else BE can not found the function

Reply:
if be can obtain length, so optimize in runtime could be used. Such as low cardinality optimization in building hash table

Reply:
so you mean this?
```java
    @Override
    public DataType getDataType() {
        if (getLength() instanceof IntegerLiteral) {
            return VarcharType.createVarcharType(((IntegerLiteral) getLength()).getValue());
        }
        return VarcharType.SYSTEM_DEFAULT;
    }
```